### ISSUE TITLE
Reference Latent Conditioning Strengh

### ISSUE THREAD

**Issue by giorgosg** [capitan01R/ComfyUI-Flux2Klein-Enhancer#6]
I was trying to figure out with claude why LatentMultipy or your implementation don't seem to work. I don't know enough to say for sure, but it seems confident it has found the way to implement it. I only have access to cloud comfy where I can't install and test my own nodes, but hopefully this makes sense to you.

[klein-ref-weight-implementation.md](https://github.com/user-attachments/files/26142555/klein-ref-weight-implementation.md)

**Comment by giorgosg** [comment]
# Per-Reference Latent Weighting — Why the Current Implementation Doesn't Work and How to Fix It

Hi, first of all thank you for the Flux2Klein-Enhancer pack — the Text/Ref Balance and Detail Controller nodes are genuinely useful. I've been digging into the ComfyUI and FLUX model source code to understand why `Ref Latent Controller` doesn't seem to have a meaningful effect, and I think I found both the problem and a correct solution.

---

## Why the current `ref_strength` scaling doesn't work

The current implementation scales the raw reference latent tensor before it enters the model:

```python
ref_latents[0] = ref_latents[0] * strength
```

The problem is what happens to that tensor inside the model (`comfy/ldm/flux/model.py`):

```python
# In _forward():
kontext, kontext_ids = self.process_img(ref)   # patch rearrange
img = torch.cat([img, kontext], dim=1)         # concatenated with main image tokens
```

Then in `forward_orig()`:
```python
img = self.img_in(img)   # Linear projection — shared across all tokens
```

And then inside every `DoubleStreamBlock`:
```python
img_modulated = self.img_norm1(img)   # ← LayerNorm here
```

**`img_norm1` is a `LayerNorm` with `elementwise_affine=False`.** This normalizes each token's values to zero mean and unit variance — completely independent of the token's original magnitude. Any scalar multiplication applied before this point is entirely erased. The 0.3x tensor and the 1.5x tensor produce identical normalized tokens.

This is why empirical testing shows `ref_strength` having no meaningful effect — the LayerNorm undoes it at every single block.

---

## Where scaling actually survives — the correct hook point

Inside `DoubleStreamBlock.forward()`, after all normalization has already happened:

```python
img_q, img_k, img_v = img_qkv.view(...).permute(2, 0, 3, 1, 4)
img_q, img_k = self.img_attn.norm(img_q, img_k, img_v)   # QKNorm — last normalization

# ... then:
if "attn1_patch" in transformer_patches:
    for p in patch:
        out = p(q, k, v, extra_options=extra_options)   # ← THIS runs post-normalization
```

The `attn1_patch` hook receives `q`, `k`, `v` **after all normalization is complete**. Scaling `k` and `v` here for specific reference token positions will genuinely affect how strongly those reference tokens influence attention — and it will not be normalized away.

`reference_image_num_tokens` is available in `extra_options`, so you know exactly which token positions belong to each reference image.

---

## Proposed implementation

A new node — `Ref Latent Weight` — that patches the model's attention using `set_model_attn1_patch`:

```python
class RefLatentWeight(io.ComfyNode):
    @classmethod
    def define_schema(cls):
        return io.Schema(
            node_id="RefLatentWeight",
            display_name="Ref Latent Weight",
            category="advanced/conditioning/edit_models",
            inputs=[
                io.Model.Input("model"),
                io.Int.Input("reference_index", default=0, min=0, max=7),
                io.Float.Input("weight", default=1.0, min=0.0, max=5.0, step=0.05),
            ],
            outputs=[
                io.Model.Output(),
            ],
        )

    @classmethod
    def execute(cls, model, reference_index, weight):
        m = model.clone()

        def ref_weight_patch(q, k, v, extra_options, **kwargs):
            ref_tokens = extra_options.get("reference_image_num_tokens", [])
            if not ref_tokens or reference_index >= len(ref_tokens):
                return {}
            total_ref = sum(ref_tokens)
            start = sum(ref_tokens[:reference_index])
            end = start + ref_tokens[reference_index]
            # Scale k and v for this specific reference's token range
            k = k.clone()
            v = v.clone()
            k[:, :, -total_ref + start : -total_ref + end] *= weight
            v[:, :, -total_ref + start : -total_ref + end] *= weight
            return {"q": q, "k": k, "v": v}

        m.set_model_attn1_patch(ref_weight_patch)
        return io.NodeOutput(m)
```

You chain one node per reference image, each with its own `reference_index` and `weight`. The `set_model_attn1_patch` infrastructure is already officially supported in ComfyUI core (`#12861`, `#12878`).

---

## Use cases this would unlock

**The primary use case — per-reference weighting in Klein multi-image workflows:**
When using 2-4 reference images, currently all references have equal influence. With per-reference weighting you could for example use one image as a strong structural reference (weight 1.0) and a second as a subtle style hint (weight 0.3).

**Upscaling with self-reference:**
A common upscaling workflow feeds the same image as both reference and latent. Being able to tune the reference weight independently of CFG gives much finer control over fidelity vs. detail enhancement — without CFG affecting both text and reference simultaneously.

**Progressive multi-reference blending:**
Weight reference A heavily during early denoising steps (structure) and reference B heavily during late steps (detail/texture) — combined with `ConditioningSetTimestepRange` on the text side this would give very precise multi-reference control.

---

## Open question

We are fairly confident this is the correct hook point, but if there is any additional normalization or projection applied to k/v inside the attention function itself (`comfy/ldm/flux/math.py`) that we haven't checked, the scaling could still be affected. You would know better than us whether `attention()` applies any further normalization to k before the softmax.

Happy to discuss further — this seems like a small implementation that would add genuinely novel functionality.

**Comment by capitan01R** [comment]
done!!
https://github.com/capitan01R/ComfyUI-Flux2Klein-Enhancer/blob/main/flux2_klein_ref_controller.py
